### PR TITLE
fix(es/transforms): Fix `this` in async class properties

### LINKED ISSUE #3228: "this" inside arrow in a static class property is compiled to (void 0)
### Describe the bug

Using `this` inside a fat arrow function on a static class property does not work, because this is compiled to `(void 0)`;

### Input code

```typescript
export class StaticTestClass {
    static testProp = 'Hello world!';

    static testMethod = () => {
        console.log(this.testProp);
    }
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": true
    },
    "target": "es2015",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "commonjs"
  },
  "minify": false,
  "isModule": true,
  "env": {
    "targets": "node 12"
  }
}
```

### Playground link

https://play.swc.rs/?version=1.2.128&code=H4sIAAAAAAAAA0utKMgvKlFIzkksLlYILkksyUwOSS0ucQbzq7kUgKAYLKpQAhQOKMovULBVUPdIzcnJVyjPL8pJUVS35kJX5ptakpGfAlSooalgawc1BgSS8%2FOK83NS9XLy0zVKMjKL9WCGalqD1dRy1QIAgrT4spEAAAA%3D&config=H4sIAAAAAAAAA0VOSwqDMBC9SnlrF1Xoxjv0EEFHieTHTCyVkLt3IhZ3j%2Fcv2GTCWJAMC3FDcoRsvhiRj0QysU0ZHbIolXmnqtjwSlkdJMOzf6nqYhTCuBgn1MHbYJejdU3RJyaRWzJhdX9n1S4f570R5ZzTTo34GDZBvYuusJX35W5HOlD4nMHzjm4gxJke%2FYBaf%2FJ6D0jYAAAA

### Expected behavior

Expected the `testMethod` method to compile to:
```
StaticTestClass.testMethod = ()=>{
    console.log(StaticTestClass.testProp);
};
```

### Actual behavior

The `testMethod` method compiles to:
```
StaticTestClass.testMethod = ()=>{
    console.log((void 0).testProp);
};
```

### Version

1.2.128

### Additional context

Only seems to happen with typescript in combination with commonjs.

### PR DESCRIPTION
**Pull request by kdy1**
swc_ecma_transforms_compat:
 - `async_to_generator`: Fix `this` in async class properties.
 - Add a test for #3228 (Closes #3228).

### CODE REVIEW

File: crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs [hunk 1]
```diff
@@ -571,47 +571,13 @@ impl Actual {
...
-                        args: vec![this.clone()],
-                        type_args: Default::default(),
-                    })
-                } else {
-                    make_fn_ref(fn_expr, None, false)
-                };
+                let fn_expr = wrapper.function.expect_fn_expr();
 
+                // We should not bind `this` in FunctionWrapper
+                wrapper.function = make_fn_ref(fn_expr, None, true);
```
Comment (magic-akari) [inline crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs:580]: we should handle `this`, `arguments`, `new.target`, `super` in function or something else by calling Hoister.

File: crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs [hunk 2]
```diff
@@ -1370,10 +1370,10 @@ let TestClass = {
     name: 'John Doe',
     testMethodFailure () {
         return new Promise(function() {
-            var _ref = _asyncToGenerator((function*(resolve) {
+            var _ref = _asyncToGenerator(function*(resolve) {
```
Comment (Austaras) [inline crates/swc_ecma_transforms_compat/tests/es2017_async_to_generator.rs:1373]: we need to pass this throught `FnEnvHoister` or something like [babel](https://babeljs.io/repl#?browsers=chrome%2050&build=&builtIns=false&corejs=3.6&spec=false&loose=true&code_lz=DYUwLgBAKiDOYGFgENawgXggbwFAQgDtkBbEALggCIApAewAtCIAROkKgGnwjDjACy4BnQAmAMWQBLYAFcATiAAUAShw8CisAuaEQAdwgAFeXRJTYy1AE9CAYwhLFsOsABuINRgB86gv4g7OkIXUAA6YDoAcyUwBgsVAG4Nf0swKCkyOlkwJzhXD04IAEYABnKklIBfSoIq3CrEoA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact&prettier=false&targets=&version=7.16.9&externalPlugins=&assumptions=%7B%22ignoreFunctionLength%22%3Atrue%7D)

File: crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs (hunk 1 above)
Comment (kdy1) [reply crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs:580]: Oh, thanks!

Comment (Austaras) [reply crates/swc_ecma_transforms_compat/src/es2017/async_to_generator.rs:580]: I'm working on Hoister, the progress is stagnating though

**Pull request closed without merging**